src/monfaction.cpp

```diff
@@ -0,0 +1,217 @@
+#include "debug.h"
+#include "monfaction.h"
+#include "json.h"
+#include <vector>
+#include <queue>
+
+std::map<std::string, mf_attitude> mf_attitude_map;
+
+std::unordered_map< mfaction_str_id, mfaction_id > faction_map;
+std::vector< monfaction > faction_list;
+
+void add_to_attitude_map( const std::set< std::string > &keys, mfaction_att_map &map,
+                          mf_attitude value );
+
+void apply_base_faction( const monfaction &base, monfaction &faction );
+
+template<>
+const monfaction &int_id<monfaction>::obj() const
+{
+    if( static_cast<size_t>( _id ) >= faction_list.size() ) {
+        debugmsg( "invalid monfaction id %d", _id );
+        return mfaction_id( 0 ).obj();
+    }
+
+    return faction_list[_id];
+}
+
+template<>
+string_id<monfaction> int_id<monfaction>::id() const
+{
+    return obj().name;
+}
+
+template<>
+int_id<monfaction> string_id<monfaction>::id() const
+{
+    const auto &iter = faction_map.find( *this );
+    if( iter == faction_map.end() ) {
+        debugmsg( "invalid monfaction id %s", c_str() );
+        return mfaction_id( 0 );
+    }
+    return iter->second;
+}
+
+template<>
+const monfaction &string_id<monfaction>::obj() const
+{
+    return id().obj();
+}
+
+template<>
+bool string_id<monfaction>::is_valid() const
+{
+    return faction_map.count( *this ) > 0;
+}
+
+template<>
+int_id<monfaction>::int_id( const string_id<monfaction> &id )
+: _id( id.id() )
+{
+}
+
+mfaction_id monfactions::get_or_add_faction( const std::string &name_arg )
```

Comment:
The id wrapper class can easily contain an invalid id, there is no problem with that. Otherwise the `is_valid` function would be pointless. The class merely signals "this is or should be an id of ..." (referring to the usage of the contains string/int).

This function is probably a bit ambiguous, but you're creating a string_id from the argument anyway, so why not let the caller do that?